datafusion/ffi/src/tests/async_provider.rs

```diff
@@ -0,0 +1,282 @@
+    }
+
+    fn schema(&self) -> Arc<Schema> {
+        super::create_test_schema()
+    }
+
+    fn table_type(&self) -> datafusion::logical_expr::TableType {
+        datafusion::logical_expr::TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        _state: &dyn Session,
+        _projection: Option<&Vec<usize>>,
+        _filters: &[Expr],
+        _limit: Option<usize>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(AsyncTestExecutionPlan::new(
+            self.batch_request.clone(),
+            self.batch_receiver.resubscribe(),
+        )))
+    }
+}
+
+impl Drop for AsyncTableProvider {
+    fn drop(&mut self) {
+        self.shutdown
+            .blocking_send(false)
+            .expect("Unable to call shutdown on spawned thread.")
+    }
+}
+
+#[derive(Debug)]
+struct AsyncTestExecutionPlan {
+    properties: datafusion::physical_plan::PlanProperties,
+    batch_request: mpsc::Sender<bool>,
+    batch_receiver: broadcast::Receiver<Option<RecordBatch>>,
+}
+
+impl AsyncTestExecutionPlan {
+    pub fn new(
+        batch_request: mpsc::Sender<bool>,
+        batch_receiver: broadcast::Receiver<Option<RecordBatch>>,
+    ) -> Self {
+        Self {
+            properties: datafusion::physical_plan::PlanProperties::new(
+                EquivalenceProperties::new(super::create_test_schema()),
+                Partitioning::UnknownPartitioning(3),
+                datafusion::physical_plan::execution_plan::EmissionType::Incremental,
+                datafusion::physical_plan::execution_plan::Boundedness::Bounded,
+            ),
+            batch_request,
+            batch_receiver,
+        }
+    }
+}
+
+impl ExecutionPlan for AsyncTestExecutionPlan {
+    fn name(&self) -> &str {
+        "async test execution plan"
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn properties(&self) -> &datafusion::physical_plan::PlanProperties {
+        &self.properties
+    }
+
+    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
+        Vec::default()
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(self)
+    }
+
+    fn execute(
+        &self,
+        _partition: usize,
+        _context: Arc<datafusion::execution::TaskContext>,
+    ) -> Result<datafusion::execution::SendableRecordBatchStream> {
+        Ok(Box::pin(AsyncTestRecordBatchStream {
+            batch_request: self.batch_request.clone(),
+            batch_receiver: self.batch_receiver.resubscribe(),
+        }))
+    }
+}
+
+impl datafusion::physical_plan::DisplayAs for AsyncTestExecutionPlan {
+    fn fmt_as(
+        &self,
+        _t: datafusion::physical_plan::DisplayFormatType,
+        _f: &mut std::fmt::Formatter,
+    ) -> std::fmt::Result {
+        // Do nothing, just a test
+        Ok(())
+    }
+}
+
+struct AsyncTestRecordBatchStream {
+    batch_request: mpsc::Sender<bool>,
```

Comment:
FWIW I think you could use a tokio::stream adapter here

Reply:
In particular I am thinking https://docs.rs/tokio-stream/latest/tokio_stream/wrappers/struct.ReceiverStream.html